File: android/app/src/main/java/com/certisgroup/cathy/reactutils/modules/crypto/AesModule.java
/*
 * Copyright (c) 2020 Certis CISCO Security Pte Ltd
 * All rights reserved.
 *
 * This software is the confidential and proprietary information of
 * Certis CISCO Security Pte Ltd. ("Confidential Information").
 * You shall not disclose such Confidential Information and shall use
 * it only in accordance with the terms of the license agreement you
 * entered into with Certis CISCO Security Pte Ltd.
 */
package com.certisgroup.cathy.reactutils.modules.crypto;

import android.util.Base64;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import java.security.GeneralSecurityException;
import java.util.HashMap;
import java.util.Map;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * RN module for AES algorithms
 *
 * @author Lingqi
 */
public class AesModule extends ReactContextBaseJavaModule {

    private static final String AES_ALGO = "AES";
    private static final Map<String, String> PADDING_MAP = paddingMap();
    private static Map<String, String> paddingMap() {
        Map<String, String> map = new HashMap<>();
        map.put("CBC", "PKCS5Padding");
        map.put("CTR", "NoPadding");
        map.put("CFB8", "NoPadding");
        return map;
    }

    public AesModule(@NonNull ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @NonNull
    @Override
    public String getName() {
        return "CryptoAES";
    }

    @ReactMethod
    public void encrypt(String mode, String iv, String key, String data, Promise promise) {
        try {
            String result = this.transform(Cipher.ENCRYPT_MODE, mode, iv, key, data);
            promise.resolve(result);
        } catch (GeneralSecurityException | IllegalArgumentException e) {
            promise.reject("AES_ENCRYPT_ERROR", e.getMessage(), e);
        }
    }

    @ReactMethod
    public void decrypt(String mode, String iv, String key, String data, Promise promise) {
        try {
            String result = this.transform(Cipher.DECRYPT_MODE, mode, iv, key, data);
            promise.resolve((result));
        } catch (GeneralSecurityException | IllegalArgumentException e) {
            promise.reject("AES_DECRYPT_ERROR", e.getMessage(), e);
        }
    }

    private String transform(int op, String mode, String iv, String key, String data)
            throws GeneralSecurityException, IllegalArgumentException {
        String padding = PADDING_MAP.get(mode);
        if (padding == null) {
            throw new IllegalArgumentException("Mode not supported");
        }
        byte[] ivBytes = Base64.decode(iv, Base64.DEFAULT);
        byte[] keyBytes = Base64.decode(key, Base64.DEFAULT);
        byte[] dataBytes = Base64.decode(data, Base64.DEFAULT);
        SecretKey secretKey = new SecretKeySpec(keyBytes, AES_ALGO);
        IvParameterSpec ivParamSpec = new IvParameterSpec(ivBytes);
        Cipher cipher = Cipher.getInstance(AES_ALGO + "/" + mode + "/" + padding);
        cipher.init(op, secretKey, ivParamSpec);
        byte[] resultBytes =  cipher.doFinal(dataBytes);
        return Base64.encodeToString(resultBytes, Base64.NO_WRAP);
    }
}

File: ios/cathy/RCTUtils/Modules/Crypto/AesModule.swift
/*
 * Copyright (c) 2020 Certis CISCO Security Pte Ltd
 * All rights reserved.
 *
 * This software is the confidential and proprietary information of
 * Certis CISCO Security Pte Ltd. ("Confidential Information").
 * You shall not disclose such Confidential Information and shall use
 * it only in accordance with the terms of the license agreement you
 * entered into with Certis CISCO Security Pte Ltd.
 */
import Foundation
import CommonCrypto

/**
 * RN module for AES algorithms
 *
 * @author Lingqi
 */
@objc(CryptoAES)
class AesModule: NSObject {
    
    private let AES_ALGO = CCAlgorithm(kCCAlgorithmAES)
    private let MODE_DICT = [
        "CBC": CCMode(kCCModeCBC),
        "CTR": CCMode(kCCModeCTR),
        "CFB8": CCMode(kCCModeCFB8)
    ]
    private let PADDING_DICT = [
        "CBC": CCPadding(ccPKCS7Padding),
        "CTR": CCPadding(ccNoPadding),
        "CFB8": CCPadding(ccNoPadding)
    ]
    private let CTR_OPTION = CCModeOptions(kCCModeOptionCTR_BE)
    
    @objc(encrypt::::::)
    func encrypt(mode: String, iv: String, key: String, data: String,
                 resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        do {
            let op = CCOperation(kCCEncrypt)
            let result = try self.transform(op: op, mode: mode, iv: iv, key: key, data: data);
            resolve(result)
        } catch let error {
            reject("AES_ENCRYPT_ERROR", error.localizedDescription, error)
        }
    }
    
    @objc(decrypt::::::)
    func decrypt(mode: String, iv: String, key: String, data: String,
                 resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        do {
            let op = CCOperation(kCCDecrypt)
            let result = try self.transform(op: op, mode: mode, iv: iv, key: key, data: data)
            resolve(result)
        } catch let error {
            reject("AES_DECRYPT_ERROR", error.localizedDescription, error)
        }
    }
    
    private func transform(op: CCOperation, mode: String, iv: String, key: String, data: String) throws -> String {
        guard let padding = PADDING_DICT[mode],
            let ivData = Data(base64Encoded: iv),
            let keyData = Data(base64Encoded: key),
            let dataBuffer = Data(base64Encoded: data) else {
                throw CCError.kParamError
        }
        let ivBytes = [UInt8](ivData),
            keyBytes = [UInt8](keyData),
            dataIn = [UInt8](dataBuffer)
        var cryptorRef: CCCryptorRef?
        var status = CCCryptorCreateWithMode(op, MODE_DICT[mode]!, AES_ALGO, padding,
                ivBytes, keyBytes, keyBytes.count, nil, 0, 0, CTR_OPTION, &cryptorRef)
        if status != kCCSuccess {
            CCCryptorRelease(cryptorRef)
            throw CCError(rawValue: status)!
        }
        let dataOutAvailable = CCCryptorGetOutputLength(cryptorRef, dataIn.count, true)
        var dataOut = [UInt8](repeating: 0, count: dataOutAvailable)
        var dataOutMoved = 0
        var totalLength = 0
        status = CCCryptorUpdate(cryptorRef, dataIn, dataIn.count,
                                 &dataOut, dataOutAvailable, &dataOutMoved)
        if status != kCCSuccess {
            CCCryptorRelease(cryptorRef)
            throw CCError(rawValue: status)!
        }
        totalLength += dataOutMoved
        status = CCCryptorFinal(cryptorRef,
                                (&dataOut + dataOutMoved) as UnsafeMutableRawPointer,
                                dataOutAvailable - dataOutMoved,
                                &dataOutMoved)
        CCCryptorRelease(cryptorRef)
        if status != kCCSuccess  {
            throw CCError(rawValue: status)!
        }
        totalLength += dataOutMoved
        let resultData = Data(bytes: dataOut, count: totalLength);
        return resultData.base64EncodedString()
    }
    
    @objc
    static func requiresMainQueueSetup() -> Bool {
        return false
    }
}

private enum CCError: CCCryptorStatus, Error {
    case kParamError     = -4300
    case kBufferTooSmall = -4301
    case kMemoryFailure  = -4302
    case kAlignmentError = -4303
    case kDecodeError    = -4304
    case kUnimplemented  = -4305
    case kOverflow       = -4306
    case kKeySizeError   = -4310
}

private let ERROR_DICT: [CCError: String] = [
    .kParamError: "Illegal parameter value",
    .kBufferTooSmall: "Insufficent buffer provided for specified operation",
    .kMemoryFailure: "Memory allocation failure",
    .kAlignmentError: "Input size was not aligned properly",
    .kDecodeError: "Input data did not decode or decrypt properly",
    .kUnimplemented: "Function not implemented for the current algorithm",
    .kOverflow: "Data size overflow",
    .kKeySizeError: "Key size not supported",
]

extension CCError: LocalizedError {
    var errorDescription: String? {
        return ERROR_DICT[self]
    }
}
